Extensions/Spine/managers/pixi-spine-atlas-manager.ts

```diff
@@ -0,0 +1,190 @@
+     *
+     * @param resources The data of resource to load.
+     */
+    getOrLoad(resourceName: string): Promise<pixi_spine.TextureAtlas> {
+      const resource = this._getAtlasResource(resourceName);
+
+      if (!resource) {
+        return Promise.reject(
+          `Unable to find atlas for resource '${resourceName}'.`
+        );
+      }
+
+      if (!this._loadingSpineAtlases.get(resource)) {
+        this._loadingSpineAtlases.set(
+          resource,
+          new Promise<pixi_spine.TextureAtlas>((resolve, reject) => {
+            const onLoad: SpineAtlasManagerRequestCallback = (
+              error,
+              content
+            ) => {
+              if (error) {
+                return reject(
+                  `Error while preloading a spine atlas resource: ${error}`
+                );
+              }
+              if (!content) {
+                return reject(
+                  `Cannot reach texture atlas for resource '${resourceName}'.`
+                );
+              }
+
+              resolve(content);
+            };
+
+            this.load(resource, onLoad);
+          })
+        );
+      }
+
+      return this._loadingSpineAtlases.get(resource)!;
+    }
+
+    /**
+     * Load specified atlas resource and pass it to callback once it is loaded.
+     *
+     * @param resources The data of resource to load.
+     * @param callback The callback to pass atlas to it once it is loaded.
+     */
+    load(
+      resource: ResourceData,
+      callback: SpineAtlasManagerRequestCallback
+    ): void {
+      const game = this._resourceLoader.getRuntimeGame();
+      const embeddedResourcesMapping = game.getEmbeddedResourcesNames(
+        resource.name
+      );
+
+      if (!embeddedResourcesMapping.length)
+        return callback(
+          new Error(`${resource.name} do not have image metadata!`)
+        );
+
+      const images = embeddedResourcesMapping.reduce<{
+        [key: string]: PIXI.Texture;
+      }>((imagesMap, embeddedResourceName) => {
+        const mappedResourceName = game.resolveEmbeddedResource(
+          resource.name,
+          embeddedResourceName
+        );
+        imagesMap[
+          embeddedResourceName
+        ] = this._imageManager.getOrLoadPIXITexture(mappedResourceName);
+
+        return imagesMap;
+      }, {});
+      const onLoad = (atlas: pixi_spine.TextureAtlas) => {
+        this._loadedSpineAtlases.set(resource, atlas);
+        callback(null, atlas);
+      };
+
+      PIXI.Assets.setPreferences({
+        preferWorkers: false,
+        crossOrigin: this._resourceLoader.checkIfCredentialsRequired(
+          resource.file
+        )
+          ? 'use-credentials'
+          : 'anonymous',
+      });
+      PIXI.Assets.add(resource.name, resource.file, { images });
+      PIXI.Assets.load<pixi_spine.TextureAtlas | string>(resource.name).then(
+        (atlas) => {
+          if (typeof atlas === 'string') {
```

Comment:
Okay, it is a little bit confusing and I hoped you will not catch it...))

As I got it :

In pixi atlas parser we pass images metadata to control loading pipeline so we execute this [block](https://github.com/pixijs/spine/blob/bda2e19cb28a0df2187451bad8ccc14f1e20da23/packages/loader-base/src/atlasLoader.ts#L66) and as images are already loaded then textures fetching is sync unlike [here](https://github.com/pixijs/spine/blob/bda2e19cb28a0df2187451bad8ccc14f1e20da23/packages/loader-base/src/atlasLoader.ts#L82). It makes texture atlas to be parsed eventually and call [resolve callback](https://github.com/pixijs/spine/blob/bda2e19cb28a0df2187451bad8ccc14f1e20da23/packages/loader-base/src/atlasLoader.ts#L62) earlier than [retval is initialized](https://github.com/pixijs/spine/blob/bda2e19cb28a0df2187451bad8ccc14f1e20da23/packages/loader-base/src/atlasLoader.ts#L70). So pixi parser returns `undefined` which will be skipped by pipeline and replaced with raw atlas text

So I covered all possible scenarios to leave everything working if this issue will be fixed (that sounds not real because `pixi-spine` is stopped in development) and after migration to `spine-pixi`

Reply:
Many thanks for the in depth explanation! I think that indeed you covered the different use cases and it's not our fault then if the output is different.

We'll add a comment and then it's good.

Reply:
@4ian I opened issue in `pixi-spine` repo and added comment with link to it in the code.
In addition I noticed that I pushed my last review changes to wrong branch. I already cherry picked them and now you can review latest updates here. I am really sorry for that.

Anyway Christmas🎄 is coming and spine extension merging is the sweetest gift of all... Just a thought🤪